### PR TITLE
[Character] Record character stats to `character_stats_record` table

### PR DESCRIPTION
**Pull request by Akkadius**
This PR creates a flat table that is used as a snapshot table to query flat data for things like

* Character leaderboards
* Magelo

Snapshots are taking when a character zones and disconnects (camps). Character save was going to be the hook point but we abuse it a bit too much in code and it doesn't need to be recorded that frequently.

Stats are unbuffed (no spells)

See table schema for fields

```sql
CREATE TABLE `character_stats_record`  (
  `character_id` int NULL,
  `name` varchar(11) NULL,
  `status` int NULL DEFAULT 0,
  `level` int NULL DEFAULT 0,
  `class` int NULL DEFAULT 0,
  `race` int NULL DEFAULT 0,
  `aa_points` int NULL DEFAULT 0,
  `hp` bigint NULL DEFAULT 0,
  `mana` bigint NULL DEFAULT 0,
  `endurance` bigint NULL DEFAULT 0,
  `ac` int NULL DEFAULT 0,
  `strength` int NULL DEFAULT 0,
  `stamina` int NULL DEFAULT 0,
  `dexterity` int NULL DEFAULT 0,
  `agility` int NULL DEFAULT 0,
  `intelligence` int NULL DEFAULT 0,
  `wisdom` int NULL DEFAULT 0,
  `charisma` int NULL DEFAULT 0,
  `magic_resist` int NULL DEFAULT 0,
  `fire_resist` int NULL DEFAULT 0,
  `cold_resist` int NULL DEFAULT 0,
  `poison_resist` int NULL DEFAULT 0,
  `disease_resist` int NULL DEFAULT 0,
  `corruption_resist` int NULL DEFAULT 0,
  `heroic_strength` int NULL DEFAULT 0,
  `heroic_stamina` int NULL DEFAULT 0,
  `heroic_dexterity` int NULL DEFAULT 0,
  `heroic_agility` int NULL DEFAULT 0,
  `heroic_intelligence` int NULL DEFAULT 0,
  `heroic_wisdom` int NULL DEFAULT 0,
  `heroic_charisma` int NULL DEFAULT 0,
  `heroic_magic_resist` int NULL DEFAULT 0,
  `heroic_fire_resist` int NULL DEFAULT 0,
  `heroic_cold_resist` int NULL DEFAULT 0,
  `heroic_poison_resist` int NULL DEFAULT 0,
  `heroic_disease_resist` int NULL DEFAULT 0,
  `heroic_corruption_resist` int NULL DEFAULT 0,
  `haste` int NULL DEFAULT 0,
  `accuracy` int NULL DEFAULT 0,
  `attack` int NULL DEFAULT 0,
  `avoidance` int NULL DEFAULT 0,
  `clairvoyance` int NULL DEFAULT 0,
  `combat_effects` int NULL DEFAULT 0,
  `damage_shield_mitigation` int NULL DEFAULT 0,
  `damage_shield` int NULL DEFAULT 0,
  `dot_shielding` int NULL DEFAULT 0,
  `hp_regen` int NULL DEFAULT 0,
  `mana_regen` int NULL DEFAULT 0,
  `endurance_regen` int NULL DEFAULT 0,
  `shielding` int NULL DEFAULT 0,
  `spell_damage` int NULL DEFAULT 0,
  `spell_shielding` int NULL DEFAULT 0,
  `strikethrough` int NULL DEFAULT 0,
  `stun_resist` int NULL DEFAULT 0,
  `backstab` int NULL DEFAULT 0,
  `wind` int NULL DEFAULT 0,
  `brass` int NULL DEFAULT 0,
  `string` int NULL DEFAULT 0,
  `percussion` int NULL DEFAULT 0,
  `singing` int NULL DEFAULT 0,
  `baking` int NULL DEFAULT 0,
  `alchemy` int NULL DEFAULT 0,
  `tailoring` int NULL DEFAULT 0,
  `blacksmithing` int NULL DEFAULT 0,
  `fletching` int NULL DEFAULT 0,
  `brewing` int NULL DEFAULT 0,
  `jewelry` int NULL DEFAULT 0,
  `pottery` int NULL DEFAULT 0,
  `research` int NULL DEFAULT 0,
  `alcohol` int NULL DEFAULT 0,
  `fishing` int NULL DEFAULT 0,
  `tinkering` int NULL DEFAULT 0,
  `created_at` datetime DEFAULT NULL,  
  `updated_at` datetime DEFAULT NULL,
  PRIMARY KEY (`character_id`)
);
```

### CODE REVIEW

**Comment by Akkadius** [comment]
With timestamps

```
MariaDB [peq]> select * from character_stats_record \G;
*************************** 1. row ***************************
            character_id: 1
                    name: Akkadius
                  status: 255
                   level: 80
                   class: 1
                    race: 2
               aa_points: 0
                      hp: 21255
                    mana: 0
               endurance: 21559
                      ac: 3505
                strength: 403
                 stamina: 440
               dexterity: 393
                 agility: 428
            intelligence: 395
                  wisdom: 372
                charisma: 383
            magic_resist: 575
             fire_resist: 575
             cold_resist: 575
           poison_resist: 575
          disease_resist: 575
       corruption_resist: 15
         heroic_strength: 48
          heroic_stamina: 85
        heroic_dexterity: 38
          heroic_agility: 73
     heroic_intelligence: 40
           heroic_wisdom: 17
         heroic_charisma: 28
     heroic_magic_resist: 0
      heroic_fire_resist: 0
      heroic_cold_resist: 0
    heroic_poison_resist: 0
   heroic_disease_resist: 0
heroic_corruption_resist: 0
                   haste: 145
                accuracy: 150
                  attack: 1835
               avoidance: 100
            clairvoyance: 87
          combat_effects: 100
damage_shield_mitigation: 15
           damage_shield: 25
           dot_shielding: 35
                hp_regen: 0
              mana_regen: 0
         endurance_regen: 0
               shielding: 35
            spell_damage: 50
         spell_shielding: 35
           strikethrough: 35
             stun_resist: 34
                backstab: 0
                    wind: 31
                   brass: 29
                  string: 32
              percussion: 0
                 singing: 0
                  baking: 300
                 alchemy: 0
               tailoring: 300
           blacksmithing: 300
               fletching: 300
                 brewing: 300
                 jewelry: 300
                 pottery: 300
                research: 0
                 alcohol: 375
                 fishing: 200
               tinkering: 300
              created_at: 2023-08-01 00:33:56
              updated_at: 2023-08-01 00:34:02
1 row in set (0.000 sec)
```

File: zone/client_packet.cpp [hunk 1]
```diff
@@ -16392,3 +16395,92 @@ void Client::Handle_OP_RaidClearNPCMarks(const EQApplicationPacket* app)
...
+		CharacterID()
+	);
+
+	r.status                   = Admin();
+	r.name                     = GetCleanName();
+	r.aa_points                = GetAAPoints() + GetSpentAA();
+	r.level                    = GetLevel();
+	r.class_                   = GetBaseClass();
+	r.race                     = GetBaseRace();
+	r.aa_points                = GetAAPoints();
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:16412]: Did the unspent part get removed?

Comment (Kinglykrab) [reply zone/client_packet.cpp:16412]: Oh nope, just duplicate information.

File: zone/client_packet.cpp [hunk 2]
```diff
@@ -16392,3 +16395,92 @@ void Client::Handle_OP_RaidClearNPCMarks(const EQApplicationPacket* app)
...
+	r.heroic_magic_resist      = GetHeroicMR() - GetSpellBonuses().HeroicMR;
+	r.heroic_fire_resist       = GetHeroicFR() - GetSpellBonuses().HeroicFR;
+	r.heroic_cold_resist       = GetHeroicCR() - GetSpellBonuses().HeroicCR;
+	r.heroic_poison_resist     = GetHeroicPR() - GetSpellBonuses().HeroicPR;
+	r.heroic_disease_resist    = GetHeroicDR() - GetSpellBonuses().HeroicDR;
+	r.heroic_corruption_resist = GetHeroicCorrup() - GetSpellBonuses().HeroicCorrup;
+	r.haste                    = GetHaste();
+	r.accuracy                 = GetAccuracy() - GetSpellBonuses().Accuracy[EQ::skills::HIGHEST_SKILL + 1];
+	r.attack                   = GetTotalATK() - GetSpellBonuses().ATK;
+	r.avoidance                = GetAvoidance();
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:16446]: Do we not have a stat bonus for avoidance?

File: zone/client_packet.cpp [hunk 3]
```diff
@@ -16392,3 +16395,92 @@ void Client::Handle_OP_RaidClearNPCMarks(const EQApplicationPacket* app)
...
+	r.clairvoyance             = GetClair() - GetSpellBonuses().Clairvoyance;
+	r.combat_effects           = GetCombatEffects();
+	r.damage_shield_mitigation = GetDSMit() - GetSpellBonuses().DSMitigation;
+	r.damage_shield            = GetDS() - GetSpellBonuses().DamageShield;
+	r.dot_shielding            = GetDoTShield() - GetSpellBonuses().DoTShielding;
+	r.hp_regen                 = GetHPRegen() - GetSpellBonuses().HPRegen;
+	r.mana_regen               = GetManaRegen() - GetSpellBonuses().ManaRegen;
+	r.endurance_regen          = GetEnduranceRegen() - GetSpellBonuses().EnduranceRegen;
+	r.shielding                = GetShielding();
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:16455]: So we not have a stay bonus for shielding?

**Comment by fryguy503** [comment]
I think Name needs to be larger?

[QueryErr] Zone [bazaar] [1406] [Data too long for column 'name' at row 1]
[INSERT INTO character_stats_record (character_id, name, status, level, class, race, aa_points, hp, mana, endurance, ac, strength, stamina, dexterity, agility, intelligence, wisdom, charisma, magic_resist, fire_resist, cold_resist, poison_resist, disease_resist, corruption_resist, heroic_strength, heroic_stamina, heroic_dexterity, heroic_agility, heroic_intelligence, heroic_wisdom, heroic_charisma, heroic_magic_resist, heroic_fire_resist, heroic_cold_resist, heroic_poison_resist, heroic_disease_resist, heroic_corruption_resist, haste, accuracy, attack, avoidance, clairvoyance, combat_effects, damage_shield_mitigation, damage_shield, dot_shielding, hp_regen, mana_regen, endurance_regen, shielding, spell_damage, spell_shielding, strikethrough, stun_resist, backstab, wind, brass, string, percussion, singing, baking, alchemy, tailoring, blacksmithing, fletching, brewing, jewelry, pottery, research, alcohol, fishing, tinkering, created_at, updated_at)  VALUES (74396,'Fundsferdiggin',0,1,16,8,0,33,0,21,10,125,95,100,41,60,83,45,30,25,25,20,15,15,0,0,0,0,0,0,0,0,0,0,0,0,0,100,0,93,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,10,0,FROM_UNIXTIME(1690896661),FROM_UNIXTIME(1690896661))]

File: zone/client_packet.cpp (hunk 1 above)
Comment (fryguy503) [reply zone/client_packet.cpp:16412]: Currently aa_points are unspent aa points, should have both, spent and unspent, possibly as separate fields?

Comment (Akkadius) [reply zone/client_packet.cpp]: They're as both a few lines above, removed the duplicate

File: zone/client_packet.cpp (hunk 2 above)
Comment (Akkadius) [reply zone/client_packet.cpp:16445]: Any that don't have it - I didn't find an equivalent for

File: zone/client_packet.cpp [hunk 4]
```diff
@@ -16392,3 +16395,91 @@ void Client::Handle_OP_RaidClearNPCMarks(const EQApplicationPacket* app)
...
+	r.dot_shielding            = GetDoTShield() - GetSpellBonuses().DoTShielding;
+	r.hp_regen                 = GetHPRegen() - GetSpellBonuses().HPRegen;
+	r.mana_regen               = GetManaRegen() - GetSpellBonuses().ManaRegen;
+	r.endurance_regen          = GetEnduranceRegen() - GetSpellBonuses().EnduranceRegen;
+	r.shielding                = GetShielding() - GetSpellBonuses().MeleeMitigation;
+	r.spell_damage             = GetSpellDmg() - GetSpellBonuses().SpellDmg;
+	r.spell_shielding          = GetSpellShield() - GetSpellBonuses().SpellShield;
+	r.strikethrough            = GetStrikeThrough() - GetSpellBonuses().StrikeThrough;
+	r.stun_resist              = GetStunResist() - GetSpellBonuses().StunResist;
+	r.backstab                 = 0;
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:16459]: What is this?

**Comment by Kinglykrab** [comment]
Works well for me, not sure what `backstab` is though.
```sql
INSERT INTO `peq`.`character_stats_record` (`character_id`, `name`, `status`, `level`, `class`, `race`, `aa_points`, `hp`, `mana`, `endurance`, `ac`, `strength`, `stamina`, `dexterity`, `agility`, `intelligence`, `wisdom`, `charisma`, `magic_resist`, `fire_resist`, `cold_resist`, `poison_resist`, `disease_resist`, `corruption_resist`, `heroic_strength`, `heroic_stamina`, `heroic_dexterity`, `heroic_agility`, `heroic_intelligence`, `heroic_wisdom`, `heroic_charisma`, `heroic_magic_resist`, `heroic_fire_resist`, `heroic_cold_resist`, `heroic_poison_resist`, `heroic_disease_resist`, `heroic_corruption_resist`, `haste`, `accuracy`, `attack`, `avoidance`, `clairvoyance`, `combat_effects`, `damage_shield_mitigation`, `damage_shield`, `dot_shielding`, `hp_regen`, `mana_regen`, `endurance_regen`, `shielding`, `spell_damage`, `spell_shielding`, `strikethrough`, `stun_resist`, `backstab`, `wind`, `brass`, `string`, `percussion`, `singing`, `baking`, `alchemy`, `tailoring`, `blacksmithing`, `fletching`, `brewing`, `jewelry`, `pottery`, `research`, `alcohol`, `fishing`, `tinkering`, `created_at`, `updated_at`) VALUES (6, 'Geki', 255, 100, 1, 522, 100, 43057, 0, 44411, 4140, 615, 622, 618, 615, 547, 552, 598, 675, 675, 675, 675, 675, 15, 160, 167, 163, 160, 92, 97, 143, 0, 0, 0, 0, 0, 0, 145, 150, 833, 100, 250, 100, 37, 30, 35, 0, 0, 0, 35, 250, 35, 35, 35, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, '2023-08-01 16:34:53', '2023-08-01 16:39:27');
```

**Review by Kinglykrab — APPROVED** [review]

**Comment by fryguy503** [comment]
> Could just be my source, but I noticed that hp_regen and mana_regen are reporting negative values in the stats table.

@Akkadius Can you confirm if the negative values are expected or if its just my source.

**Comment by Akkadius** [comment]
> > Could just be my source, but I noticed that hp_regen and mana_regen are reporting negative values in the stats table.
> 
> @Akkadius Can you confirm if the negative values are expected or if its just my source.

I didn't get negative numbers in my example above. So not sure offhand

File: zone/client_packet.cpp (hunk 4 above)
Comment (Akkadius) [reply zone/client_packet.cpp:16459]: Backstab is a mod - that apparently doesn't have any getters

Comment (Kinglykrab) [reply zone/client_packet.cpp:16459]: > Backstab is a mod - that apparently doesn't have any getters

👍

**Comment by Kinglykrab** [comment]
Was thinking of adding a getter for Backstab to this, just not sure which one we're needing.
image

**Pull request merged**